lib/spf/macro_string.rb

```diff
@@ -43,9 +46,120 @@ def expand(context = nil)
       return (@expanded = @text) unless @text =~ /%/
         # Short-circuit expansion if text has no '%' characters.
 
+      server, request = context ? context : [@server, @request]
+
+      valid_context(true, server, request)
+
       expanded = ''
-      # TODO
-      return (@expanded = @text)
+
+      text = @text
+
+      while m = text.match(/ (.*?) %(.) /x) do
+        expanded += m[1]
+        key = m[2]
+
+        if (key == '{')
+          if m2 = m.post_match.match(/ (\w|_\p{Alpha}+) ([0-9]+)? (r)? ([.\-+,\/_=])? } /x)
+            char, rh_parts, reverse, delimiter = m2.captures
+
+            # Upper-case macro chars trigger URL-escaping AKA percent-encoding
+            # (RFC 4408, 8.1/26):
+            do_percent_encode = char =~ /\p{Upper}/
+            char.downcase!
+
+            if char == 's' # RFC 4408, 8.1/19
+              value = request.identity
+            elsif char == 'l' # RFC 4408, 8.1/19
+              value = request.localpart
+            elsif char == 'o' # RFC 4408, 8.1/19
+              value = request.domain
+            elsif char == 'd' # RFC 4408, 8.1/6/4
+              value = request.authority_domain
+            elsif char == 'i' # RFC 4408, 8.1/20, 8.1/21
+              ip_address = request.ip_address
+              ip_address = SPF::Util.ipv6_address_to_ipv4(ip_address) if SPF::Util.ipv6_address_is_ipv4_mapped(ip_address)
+              if IP::V4 === ip_address
+                value = ip_address.to_addr
+              elsif IP::V6 === ip_address
+                value = ip_address.to_hex.upcase.split('').join('.')
+              else
+                server.throw_result(:permerror, request, "Unexpected IP address version in request")
+              end
+            elsif char == 'p' # RFC 4408, 8.1/22
+              # RFC 7208 says this should not be used
```

Comment:
True, RFC 7208 says that, but we should still implement this. At least I'd replace it with the literal macro expression (to effectively leave it unexpanded) rather than just replacing it with an empty string, or, worse, whatever the previous value of the `value` variable is.

Reply:
Good point. I'll implement this by leaving the expression unexpanded.